### ISSUE TITLE
adding coverage part and having extension method in UndefinedObject ends with 100 GB/mo more memory Image

### ISSUE THREAD

**Issue by SabineMa** [hpi-swa/smalltalkCI#476]
With this smalltalk.ston

```
SmalltalkCISpec {
  #name : 'Spesenfuchs Tests',
  #loading : [
    SCIMetacelloLoadSpec {
  	#baseline : 'RKA24',
 	#directory : 'packages',
 	#ignoreImage : true,
 	#onConflict : #useIncoming,
	#onUpgrade : #useIncoming,
 	#onWarningLog : true,
 	#platforms : [ #pharo],
 	#useLatestMetacello : false
}
  ],
  #testing : {
    #classes : [ #RKANumberExtensionTest, #RKAStringTest],
    #defaultTimeout : 30, 
    #failOnSCIDeprecationWarnings : false,
    #failOnZeroTests : false, 
    #hidePassingTests : false, 
    #serializeError: true,
  }
}
```

My tests result in:

![Bildschirmfoto 2020-09-30 um 21 12 34](https://user-images.githubusercontent.com/13763544/94731466-1d1dbb00-0365-11eb-9e78-a1839493c5d8.png)

this is what I expect

Now I add the coverage part like this:

```
SmalltalkCISpec {
  #name : 'Spesenfuchs Tests',
  #loading : [
    SCIMetacelloLoadSpec {
  	#baseline : 'RKA24',
 	#directory : 'packages',
 	#ignoreImage : true,
 	#onConflict : #useIncoming,
	#onUpgrade : #useIncoming,
 	#onWarningLog : true,
 	#platforms : [ #pharo],
 	#useLatestMetacello : false
}
  ],
  #testing : {
    #classes : [ #RKANumberExtensionTest, #RKAStringTest],
    #defaultTimeout : 30, 
    #failOnSCIDeprecationWarnings : false,
    #failOnZeroTests : false, 
    #hidePassingTests : false, 
    #serializeError: true,
    #coverage : {
      #packages : [ 'RKA24-Model' ],
      #format : #coveralls
    }
  }
}
```

This leads to a never ending story with no more memory on my mac:

 
![Bildschirmfoto 2020-09-30 um 20 41 19](https://user-images.githubusercontent.com/13763544/94731616-60782980-0365-11eb-996c-8fc987d2be69.png)

I do not get what is wrong and it would be nice if StCi would tell me :-)

**Comment by SabineMa** [comment]
after debugging in it, I see that the problem is here:
![Bildschirmfoto 2020-09-30 um 22 00 41](https://user-images.githubusercontent.com/13763544/94733743-7dfac280-0368-11eb-9bd8-1d8bb69c1932.png)
This seems to be to many classes (102)/methods(3390)?

**Comment by MrModder** [comment]
Hm, I don't see how 3000 classes should produce anything close to 100GB in terms of memory usage...
There seems to be a serious memory leak somewhere.

I've heard from @fniephaus that there can be endless-recursion problems with CoverageWrapper objects in some cases.
It might be that only a few of the methods are actually causing problems, but with so many classes it's pretty hard to figure out where the bug originates from.
Do you have sub-packages, that you could try enabling/disabling coverage for to narrow down the problem?

**Comment by fniephaus** [comment]
As @MrModder pointed out correctly, code coverage can cause problems and it's a good idea to find out which package is causing them. Also, I wonder, what is the canonical way of determining code coverage in Pharo? Does the Test Runner still support it, how is coverage done in Pharo these days, and does that work fine for your packages, @SabineMa?

**Comment by SabineMa** [comment]
Hi @MrModder and @fniephaus, it is not 3000 classes, "only" 3000 methods. I started to narrow down the problem today. 

When removing the #packages part ([ 'RKA24-Model' ]) and exchanging it with  #classes and adding all classes of the original #packages statement (all model classes) without those classes, which are overwritten, the problem disappears. 

The classes part which works is like this:
#coverage : {    
 #classes : [#RKAAccount, #RKAAccountingRecord, #RKAClosedBetaPerson, #RKACompany, #RKACountry, #RKACurrency, #RKACurrencyCountry, #RKACustomerPerson, #RKADay, #RKAError, #RKAExchangeRate, #RKAExpensesCountry, #RKAExpensesCountryInterval, #RKAExportSetting, #RKAExportStub, #RKALogEntry, #RKALogin, #RKAPaymentInKind, #RKAPaymentMethod, #RKAPerson, #RKAPlace, #RKAPlan, #RKAPlatformNews, #RKAReceipt, #RKAReceiptKind, #RKAReceiptVat, #RKASystemMessage, #RKATrip, #RKAVat, #RKAVehicleKind, #RKAVehicleTrip, #RKAExportSettingKontolino, #RKAExportSettingSevdesk, #RKAMoney, #RKATestTrip, #ManifestRKAModel ],     
 #format : #lcov

so, for me this would be ok.

question: do you want me to investigate further for fixing the problem in StCi?
@fniephaus when I was trying test runner the last time it did not work and I stopped using it. I tried again this evening and I have the same problem, 100% cpu and increasing memory.

**Comment by gcotelli** [comment]
If it's working adding the classes maybe is some problem with an extension?
Sabine do you have extension methods in this package?
We have a project with more than 30.000 classes and the coverage is working
there, so I don't think it is a problem of code size.

**Comment by SabineMa** [comment]
@gcotelli yes, I have several extension methods. I identified 2 problems in my code:

1.) one extension method was calling a method which is not defined. I deleted it because i don't need it. 
2.) another extension method in UndefinedObject. This is a method I need, it is very simple:

```
UndefinedObject >>rkaSortString
	^ ''
```

When deleting this method, everything is find and the coverage works. But I don't want to miss this method. I tried with another method:

```
UndefinedObject >>someMethod
	^42
```

same result: hanging image. I tried this several times and it can be reproduced.

**My conclusion is, that adding extension methods to UndefinedObject causes the problem.**

@MrModder @fniephaus can you check this? Is there sth more I can do?

**Comment by gcotelli** [comment]
Maybe we can check that the coverage wrappers are installed only in the extension methods not on the whole class. Because it looks strange to me that the problem is caused by this kind of extension, but if we are wrapping methods like `ifNil:` this can be problematic.

**Comment by SabineMa** [comment]
Can I do sth here? @fniephaus @MrModder

**Comment by fniephaus** [comment]
@MrModder, @gcotelli, or I need to find some time to look into this in more detail (I'm currently busy preparing the upcoming semester). Contributions and pull requests are always welcome!

BTW: In https://github.com/hpi-swa/smalltalkCI/issues/476#issuecomment-706761143, you mentioned:
> when I was trying test runner in Pharo the last time it did not work and I stopped using it. I tried again this evening and I have the same problem, 100% cpu and increasing memory.

Have you talked to the Pharo community about this in the meantime? Were they able to find out more?

**Comment by SabineMa** [comment]
Hi @fniephaus ,

no I did not investigate further. I will do and tell you here. 
Thank you for your reaction and for your time.

**Comment by jbrichau** [comment]
@SabineMa The fix in #585 should be the fix for this issue as well.

**Comment by SabineMa** [comment]
Thank you @jbrichau 
I did not test it today because I don’t have time but seems to be ok according to the other issue.  Already read about it today. 
Currently I will not change our nightly run but when I have time I will adopt it. Then I will see.
